### PR TITLE
gh-86457: Fix signature for code.replace().

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
https://bugs.python.org/issue42291

#86457

* Issue: gh-86457

### CODE REVIEW

**Review by iritkatriel — DISMISSED** [review]
This has merge conflicts now.

File: Objects/codeobject.c [hunk 1]
```diff
@@ -621,22 +621,22 @@ code_sizeof(PyCodeObject *co, PyObject *Py_UNUSED(args))
...
-    co_code: PyBytesObject(c_default="(PyBytesObject *)self->co_code") = None
-    co_consts: object(subclass_of="&PyTuple_Type", c_default="self->co_consts") = None
-    co_names: object(subclass_of="&PyTuple_Type", c_default="self->co_names") = None
-    co_varnames: object(subclass_of="&PyTuple_Type", c_default="self->co_varnames") = None
-    co_freevars: object(subclass_of="&PyTuple_Type", c_default="self->co_freevars") = None
-    co_cellvars: object(subclass_of="&PyTuple_Type", c_default="self->co_cellvars") = None
-    co_filename: unicode(c_default="self->co_filename") = None
-    co_name: unicode(c_default="self->co_name") = None
-    co_lnotab: PyBytesObject(c_default="(PyBytesObject *)self->co_lnotab") = None
+    co_argcount: int(c_default="self->co_argcount") = unknown
```
Comment (iritkatriel) [inline Objects/codeobject.c]: Maybe "unchanged" instead of "unknown"?

Comment (serhiy-storchaka) [reply Objects/codeobject.c]: Done.

File: Tools/clinic/clinic.py [hunk 2]
```diff
@@ -2652,7 +2652,7 @@ def __init__(self, name, py_name, function, default=unspecified, *, c_default=No
         self.name = ensure_legal_c_identifier(name)
         self.py_name = py_name
 
-        if default is not unspecified:
+        if default is not unspecified and (default is not NULL or c_default == 'NULL'):
```
Comment (iritkatriel) [inline Tools/clinic/clinic.py]: What does this change do? Is it related to the signature?

**Review by iritkatriel — APPROVED** [review]
I don't understand the clinic code, but the rest LGTM.

**Comment by serhiy-storchaka** [comment]
There was a reason why it was in the draft mode. While it prevent generation of incorrect signature, it generates signature incompatible with the `inspect` module. The changes in the Argument Clinix were my attempt to generate supported signature. But it was not perfect solution. I need to work on it more.

**Review by iritkatriel — DISMISSED** [review]
This has merge conflicts now.

File: Tools/clinic/clinic.py (hunk 2 above)
Comment (serhiy-storchaka) [reply Tools/clinic/clinic.py]: Seems it is a leftover from my previous attempts. It does not affect current output, although there may be cases in which it makes difference. But for now I remove this change.

**Review by serhiy-storchaka — COMMENTED** [review]
Altering representation for separate parameters does not help much. It prevents pydoc from showing invalid signature, but does not help with showing the correct one.

Using var-keyword parameter is perfect for this. But AC does not support it. We need to override generated signature. I introduce a new directive `@text_signature` which allows to override the generated signature. @erlend-aasland, could you please look at it?

**Comment by erlend-aasland** [comment]
I will have a look, but it may take some days before I find time. Thanks for the ping, @serhiy-storchaka

File: Tools/clinic/clinic.py [hunk 3]
```diff
@@ -5489,142 +5495,145 @@ def format_docstring(self) -> str:
...
-            for p in reversed(parameters):
-                if not p.converter.show_in_signature:
-                    continue
-                if p.is_positional_only():
-                    need_a_trailing_slash = True
-                break
+        if self.forced_text_signature:
+            add(self.forced_text_signature)
+        else:
+            add('(')
```
Comment (serhiy-storchaka) [inline Tools/clinic/clinic.py:5501]: All following lines just increase indentation. GitHub is not good in showing such case.

**Comment by erlend-aasland** [comment]
It would be better if Argument Clinic could handle this by itself, without the need for the `@text_signature` override. However, this is a good workaround. It also unblocks cases in the code base where we could not apply clinic because of signature issues.

File: Tools/clinic/clinic.py (hunk 3 above)
Comment (erlend-aasland) [reply Tools/clinic/clinic.py:5501]: For the record, I plan to refactor the signature generation into a helper function. See #107623

**Comment by serhiy-storchaka** [comment]
I am sorry for creating conflicts with your PR, but this PR was planned for backporting, and your refactoring seems is not purposed for backporting.

**Comment by erlend-aasland** [comment]
> I am sorry for creating conflicts with your PR, but this PR was planned for backporting, and your refactoring seems is not purposed for backporting.

No worries; I'll handle that :)

**Pull request merged**